lib/test.rb

```diff
@@ -413,36 +435,60 @@ def setup_formulae_deps_instances(formula, formula_name)
         return
       end
 
-      build_dependents_from_source = %w[
-        cabal-install
-        ghc
-        go
-        ocaml
-        rust
-      ].include?(formula_name)
-
       uses_args = []
       uses_args << "--recursive" unless Homebrew.args.skip_recursive_dependents?
-      # whitelist specific formula where we want to test build dependants.
-      uses_args << "--include-build" if build_dependents_from_source
       dependents =
-        Utils.popen_read("brew", "uses", "--include-test", *uses_args, formula_name)
+        Utils.popen_read("brew", "uses", "--include-build", "--include-test", *uses_args, formula_name)
              .split("\n")
       dependents -= @formulae
       dependents = dependents.map { |d| Formulary.factory(d) }
 
-      if build_dependents_from_source
-        @source_dependents = dependents
-        @testable_dependents = @source_dependents.select(&:test_defined?)
-        @bottled_dependents = []
-        return
+      # Firstly, split into dependents that we could potentially be building from source
+      # and those we should not. The criteria is that it depends on a formula in the whitelist
+      # and that formula has been, or will be, built in this test run.
+      @source_dependents, dependents = dependents.partition do |f|
+        deps = if Homebrew.args.skip_recursive_dependents?
+          f.deps
+        else
+          f.recursive_dependencies
+        end
+        deps.reject!(&:optional?)
```

Comment:
Looks like a bunch of logic above here could be used in a way that doesn't require their reputation below?

Reply:
Yeah, I feel so. I suppose I could maintain a list of dependencies of the dependents (or zip the two together).